### DISCUSSION TITLE
JumpList and PublishAOT support ?

### DISCUSSION THREAD

**Discussion by latop2604** [dahall/Vanara#440] [Q&A]
Hello, I have an Avalonia app build using PublishAOT. 

And I would like to add JumpList support to it.

Using the following code snippet it work, but when running the AOT version the app crash with an `InvalidProgramException`.

Do you know is AOT is supposed to work with Vanara lib. And is yes, any idea on why it does not work?

```
Unhandled Exception: System.InvalidProgramException: Common Language Runtime detected an invalid program. The body of method 'Void CDestinationList..ctor()' is invalid.
   at Internal.Runtime.TypeLoaderExceptionHelper.CreateInvalidProgramException(ExceptionStringID, String) + 0x40
   at Internal.Runtime.CompilerHelpers.ThrowHelpers.ThrowInvalidProgramExceptionWithArgument(ExceptionStringID, String) + 0x9
   at Vanara.PInvoke.Shell32.CDestinationList..ctor() + 0x15
   at Vanara.Windows.Shell.JumpList.ApplySettings(String) + 0x38
   at MyApp.JumpLists.Init() + 0x288
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.SetupCore(String[]) + 0x46
   at Avalonia.Controls.ApplicationLifetimes.ClassicDesktopStyleApplicationLifetime.Start(String[]) + 0x12
   at MyApp!<BaseAddress>+0x14c4cae
```

```csharp
// BuiltInComInteropSupport is set to true

var jumpList = new Vanara.Windows.Shell.JumpList();

foreach (strin profile in profiles)
{
    jumpList.Add(new JumpListTask(
        profile ?? "Default",
        "C:\\myAppPath\\myApp.exe"
        )
    {
        Arguments = profile ?? "",
        Description = $"Open with profile {profile ?? "Default"}",
    });
}

jumpList.ApplySettings();
```

**Comment by latop2604** [comment]
I was able to have something working with the help of C# Discord server.
The source code is available as gist here https://gist.github.com/latop2604/73bd7d8008e7ed925f5713fda9201ced

It Use ComWrapper and [ComSourceGenerator](https://learn.microsoft.com/en-us/dotnet/standard/native-interop/comwrappers-source-generation). It's strongly based on Vanara source code but with lot of adaptation (that I don't completely understand)

It support only adding Task item to the jump list.

**Comment by dahall** [comment]
Are you using .NET 7 or 8? It appears to be failing on the constructor of the `ICustomDestinationList` constructor which calls CoCreateInstance by .NET. I'm afraid you may be running into an incompatibility between an interop library and AOT.

**Reply by latop2604** [reply]
Net8

**Comment by dahall** [comment]
Cool use of ComWrappers. I have been meaning to try this out and see if it is backwards compatible.
